### ISSUE TITLE
Add Armadillo

### ISSUE THREAD

**Issue by usernameiwantedwasalreadytaken** [Microsoft/vcpkg#1491]
It would be great if we had [armadillo](http://arma.sourceforge.net/).

[Github code here](https://github.com/conradsnicta/armadillo-code)

**Comment by Voskrese** [comment]
как пример
`git clone https://github.com/conradsnicta/armadillo-code`
`edit armadillo-code/include/armadillo_bits/config.hpp`

добавить строки и естественно сохранить
```
#define ARMA_USE_LAPACK
#define ARMA_USE_BLAS
#define ARMA_USE_ARPACK
#define ARMA_USE_SUPERLU
```
`md armadillo-code/build && cd armadillo-code/build`
`cmake .. -DCMAKE_PREFIX_PATH="E:/tools/vcpkg/installed/x86-windows" -G"Visual Studio 15 2017" -DCMAKE_CONFIGURATION_TYPES="Release" -L -DCMAKE_INSTALL_PREFIX:PATH="../_Rel" -DDETECT_HDF5=ON -DHDF5_ROOT="E:/tools/vcpkg/installed/x86-windows"`

```
-- Configuring Armadillo 7.959.0
--
-- *** WARNING: automatic installation is experimental for this platform.
-- *** WARNING: if anything breaks, you get to keep all the pieces.
-- *** WARNING: manual installation is described in README.txt
--
-- CMAKE_SYSTEM_NAME          = Windows
-- CMAKE_CXX_COMPILER_ID      = MSVC
-- CMAKE_CXX_COMPILER_VERSION = 19.11.25303.0
-- CMAKE_COMPILER_IS_GNUCXX   =
-- BUILD_SHARED_LIBS          = OFF
-- DETECT_HDF5                = ON
-- Found MKL libraries: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2017.2.187/windows/mkl/lib/ia32_win/mkl_rt.lib
-- Found OpenBLAS: E:/tools/vcpkg/installed/x86-windows/lib/openblas.lib
-- Found BLAS: E:/tools/vcpkg/installed/x86-windows/lib/BLAS.lib
-- Found LAPACK: E:/tools/vcpkg/installed/x86-windows/lib/lapack.lib
--      MKL_FOUND = YES
--   ACMLMP_FOUND = NO
--     ACML_FOUND = NO
-- OpenBLAS_FOUND = YES
--    ATLAS_FOUND = NO
--     BLAS_FOUND = YES
--   LAPACK_FOUND = YES
--
-- *** If the MKL or ACML libraries are installed in non-standard locations such as
-- *** /opt/intel/mkl, /opt/intel/composerxe/, /usr/local/intel/mkl
-- *** make sure the run-time linker can find them.
-- *** On Linux systems this can be done by editing /etc/ld.so.conf
-- *** or modifying the LD_LIBRARY_PATH environment variable.
--
-- *** On systems with SELinux enabled (eg. Fedora, RHEL),
-- *** you may need to change the SELinux type of all MKL/ACML libraries
-- *** to fix permission problems that may occur during run-time.
-- *** See README.txt for more information
--
-- HDF5_FOUND = TRUE
-- Set ARMA_HDF5_INCLUDE_DIR to E:/tools/vcpkg/installed/x86-windows/include
-- ARPACK_FOUND = NO
-- Looking for SuperLU version 5
-- Could not find SuperLU
-- SuperLU_FOUND = NO
--
-- *** Armadillo wrapper library will use the following libraries:
-- *** ARMA_LIBS = C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2017.2.187/windows/mkl/lib/ia32_win/mkl_rt.lib;E:/tools/vcpkg/installed/x86-windows/lib/hdf5.lib
--
-- Copying E:/downloads/tmp/armadillo-code-unstable/include/ to E:/downloads/tmp/armadillo-code-unstable/build/tmp/include/
-- Generating E:/downloads/tmp/armadillo-code-unstable/build/tmp/include/config.hpp
-- Generating E:/downloads/tmp/armadillo-code-unstable/examples/Makefile
-- CMAKE_CXX_FLAGS           = /DWIN32 /D_WINDOWS /W3 /GR /EHsc
-- CMAKE_SHARED_LINKER_FLAGS = /machine:X86
-- CMAKE_REQUIRED_INCLUDES   = E:/tools/vcpkg/installed/x86-windows/include
--
-- CMAKE_INSTALL_PREFIX     = ../_Rel
-- CMAKE_INSTALL_LIBDIR     = lib
-- CMAKE_INSTALL_INCLUDEDIR = include
-- CMAKE_INSTALL_DATADIR    = share
-- CMAKE_INSTALL_BINDIR     = bin
-- Generating 'E:/downloads/tmp/armadillo-code-unstable/build/ArmadilloConfig.cmake'
-- Generating 'E:/downloads/tmp/armadillo-code-unstable/build/ArmadilloConfigVersion.cmake'
-- Generating 'E:/downloads/tmp/armadillo-code-unstable/build/InstallFiles/ArmadilloConfig.cmake'
-- Generating 'E:/downloads/tmp/armadillo-code-unstable/build/InstallFiles/ArmadilloConfigVersion.cmake'
-- Copying E:/downloads/tmp/armadillo-code-unstable/misc/ to E:/downloads/tmp/armadillo-code-unstable/build/tmp/misc/
-- Generating 'E:/downloads/tmp/armadillo-code-unstable/build/tmp/misc/armadillo.pc'
-- Configuring done
-- Generating done
-- Build files have been written to: E:/downloads/tmp/armadillo-code-unstable/build
-- Cache values
ARPACK_LIBRARY:FILEPATH=ARPACK_LIBRARY-NOTFOUND
ATLAS_CBLAS_INCLUDE_DIR:PATH=E:/tools/vcpkg/installed/x86-windows/include
ATLAS_CLAPACK_INCLUDE_DIR:PATH=E:/tools/vcpkg/installed/x86-windows/include
BUILD_SHARED_LIBS:BOOL=OFF
CMAKE_INSTALL_PREFIX:PATH=../_Rel
DETECT_HDF5:BOOL=ON
HDF5_DIR:PATH=E:/tools/vcpkg/installed/x86-windows/share/hdf5
INSTALL_BIN_DIR:STRING=
INSTALL_DATA_DIR:STRING=
INSTALL_INCLUDE_DIR:STRING=
INSTALL_LIB_DIR:STRING=
MKL_ROOT:STRING=C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2017.2.187\windows\mkl
SuperLU_INCLUDE_DIR:PATH=SuperLU_INCLUDE_DIR-NOTFOUND
SuperLU_LIBRARY:FILEPATH=SuperLU_LIBRARY-NOTFOUND
atlas_LIBRARY:FILEPATH=atlas_LIBRARY-NOTFOUND
mkl_rt_LIBRARY:FILEPATH=C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2017.2.187/windows/mkl/lib/ia32_win/mkl_rt.lib
openblas_LIBRARY:FILEPATH=E:/tools/vcpkg/installed/x86-windows/lib/openblas.lib
openblaso_LIBRARY:FILEPATH=openblaso_LIBRARY-NOTFOUND
openblasp_LIBRARY:FILEPATH=openblasp_LIBRARY-NOTFOUND
satlas_LIBRARY:FILEPATH=satlas_LIBRARY-NOTFOUND
tatlas_LIBRARY:FILEPATH=tatlas_LIBRARY-NOTFOUND
```
`cmake --build .`
build ok )))
`cmake -P cmake_install.cmake`

**depens NOTFOUND**
`SuperLU__FOUND, ACMLMP_FOUND, ACML_FOUND, ATLAS_FOUND, openblaso_LIBRARY-NOTFOUND, openblasp_LIBRARY-NOTFOUND, satlas_LIBRARY-NOTFOUND, tatlas_LIBRARY-NOTFOUND`

**Comment by usernameiwantedwasalreadytaken** [comment]
Installing dependencies:

```
vcpkg install openblas clapack hdf5
```

Configuring (with no changes to `armadillo-code/include/armadillo_bits/config.hpp`):

```
cmake .. -DCMAKE_PREFIX_PATH="C:/tools/vcpkg/installed/x64-windows" -G"Visual Studio 15 2017 Win64" -DCMAKE_CONFIGURATION_TYPES="Release x64" -L -DCMAKE_INSTALL_PREFIX:PATH="../_Rel" -DDETECT_HDF5=ON -DHDF5_ROOT="C:/tools/vcpkg/installed/x64-windows"
```

Building:

```
cmake --build . --config "Release x64"
```

Works.

P.S.: We would still miss SuperLU and Arpack support. I think to add [SuperLU](http://crd-legacy.lbl.gov/~xiaoye/SuperLU/#superlu) to vcpkg would not be difficult, though. And [Arpack](http://www.caam.rice.edu/software/ARPACK/) is (in my opinion) less critical to use Armadillo...

**Comment by usernameiwantedwasalreadytaken** [comment]
As for SuperLU,

```
git clone https://github.com/xiaoyeli/superlu
mkdir superlu\build && cd superlu\build
cmake .. -DXSDK_ENABLE_Fortran=FALSE -Denable_tests=FALSE -DCMAKE_PREFIX_PATH="C:/tools/vcpkg/installed/x64-windows" -G"Visual Studio 15 2017 Win64" -DCMAKE_CONFIGURATION_TYPES="Release" -L -DCMAKE_INSTALL_PREFIX:PATH="../_Rel"
cmake --build . --config "Release"
```

Works too. 

But specifying `-DBUILD_SHARED_LIBS=TRUE` on the CMake command line to build a dynamic library doesn't work.

**Comment by usernameiwantedwasalreadytaken** [comment]
Building as a dynamic library works if we use openblas:

```
cmake .. -DBUILD_SHARED_LIBS=TRUE -DXSDK_ENABLE_Fortran=FALSE -Denable_tests=FALSE -Denable_blaslib=FALSE -DTPL_BLAS_LIBRARIES="C:/tools/vcpkg/installed/x64-windows/lib/openblas.lib" -DCMAKE_PREFIX_PATH="C:/tools/vcpkg/installed/x64-windows" -G"Visual Studio 15 2017 Win64" -DCMAKE_CONFIGURATION_TYPES="Release" -L -DCMAKE_INSTALL_PREFIX:PATH="../_Rel"
```

But before we need to manually patch `SRC\CMakeLists.txt` changing line 236 from this:

```
target_link_libraries(superlu ${BLAS_LIB} m)
```

into this:

```
target_link_libraries(superlu ${BLAS_LIB})
```

**Comment by ShannonZ** [comment]
Only support x86 editions? I found no x64 lib installed

**Comment by arthur-tacca** [comment]
I believe this should be closed due to #2954?
